sensor-gateway/common/error/SensorAccessLinkError.cpp

```diff
@@ -0,0 +1,161 @@
+                                             Severity const& severity,
+                                             ErrorCode const& code,
+                                             std::string const& message,
+                                             HighResolutionTimePoint timestamp) :
+        runtime_error(buildDetailedMessage(timestamp, origin, category, severity, code, message)),
+        timestamp(timestamp),
+        origin(origin),
+        category(category),
+        severity(severity),
+        code(code),
+        message(message) {}
+
+SensorAccessLinkError::~SensorAccessLinkError() noexcept {}
+
+SensorAccessLinkError::SensorAccessLinkError(SensorAccessLinkError const& other) noexcept :
+        SensorAccessLinkError(other.origin, other.category, other.severity, other.code, other.message,
+                              other.timestamp) {
+}
+
+SensorAccessLinkError::SensorAccessLinkError(SensorAccessLinkError&& other) noexcept :
+        SensorAccessLinkError(std::move(other.origin), other.category, other.severity, other.code,
+                              std::move(other.message), std::move(other.timestamp)) {
+}
+
+SensorAccessLinkError& SensorAccessLinkError::operator=(SensorAccessLinkError const& other)& {
+    SensorAccessLinkError temporary(other);
+    swap(*this, temporary);
+    return *this;
+}
+
+SensorAccessLinkError& SensorAccessLinkError::operator=(SensorAccessLinkError&& other)& noexcept {
+    swap(*this, other);
+    return *this;
+}
+
+void SensorAccessLinkError::swap(SensorAccessLinkError& current, SensorAccessLinkError& other) noexcept {
+    std::swap(current.origin, other.origin);
+    std::swap(current.category, other.category);
+    std::swap(current.severity, other.severity);
+    std::swap(current.code, other.code);
+    std::swap(current.message, other.message);
+    std::swap(current.timestamp, other.timestamp);
+}
+
+bool SensorAccessLinkError::operator==(SensorAccessLinkError const& other) const {
+    auto sameOrigin = (origin == other.origin);
+    auto sameCategory = (category == other.category);
+    auto sameSeverity = (severity == other.severity);
+    auto sameCode = (code == other.code);
+    auto sameMessage = (message == other.message);
+    auto messagesAreEqual = (sameOrigin &&
+                             sameCategory &&
+                             sameSeverity &&
+                             sameCode &&
+                             sameMessage
+    );
+    return messagesAreEqual;
+}
+
+bool SensorAccessLinkError::operator!=(SensorAccessLinkError const& other) const {
+    return !operator==(other);
+}
+
+SensorAccessLinkError const SensorAccessLinkError::returnDefaultData() noexcept {
+    return Defaults::DEFAULT_SENSOR_ACCESS_LINK_ERROR;
+}
+
+bool SensorAccessLinkError::isFatalForGateway() const noexcept {
+    bool isFatal = false;
+    if (severity == EMERGENCY) {
+        isFatal = true;
+    }
+    return isFatal;
```

Comment:
Nope, it's neither more explicit nor more or less optimal.  
I guaranty than the compiled and executed code **IS NOT** a textual representation of what is written in both the actual code and the alternative solution you suggested.   
If anything it is more likely the actual code reduces the pre-compiler work.  
In any case, "performance" and "optimalness" are of no concern here.  
And the current code is cleaner and more explicit because of one return statement and one declaration